File: collector/src/main/java/com/bitmovin/analytics/utils/HttpClient.java
package com.bitmovin.analytics.utils;

import android.content.Context;
import android.util.Log;

import com.bitmovin.analytics.BitmovinAnalyticsConfig;
import com.bitmovin.analytics.CollectorConfig;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class HttpClient {
    public static final MediaType JSON
            = MediaType.parse("application/json; charset=utf-8");
    private static final String TAG = "HttpClient";
    private OkHttpClient client = null;
    private Context context;
    private boolean tryResendDataOnFailedConnection;

    public HttpClient(Context context, boolean tryResendDataOnFailedConnection) {
        this.context = context;
        this.tryResendDataOnFailedConnection = tryResendDataOnFailedConnection;
    }


    public void post(String url, String postBody, final Callback callback) {
        if (client == null) {
            if (tryResendDataOnFailedConnection){
                Log.d("RetryBackend", "retry");
                client = new OkHttpClient.Builder()
                        .retryOnConnectionFailure(false).build();
            } else {
                Log.d("RetryBackend", "regular");
                client = new OkHttpClient();
            }
        }

        Log.d(TAG, String.format("Posting Analytics JSON: \n%s\n", postBody));
        Request request = new Request.Builder()
                .url(url)
                .header("Origin", String.format("http://%s", context.getPackageName()))
                .post(RequestBody.create(JSON, postBody))
                .build();


        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                Log.e(TAG, "HTTP Error: ", e);
                if (callback != null) {
                    callback.onFailure(call, e);
                }
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (callback != null) {
                    callback.onResponse(call, response);
                }
                if (response == null) {
                    return;
                }
                response.close();
            }


        });
    }
}

File: collector/src/main/java/com/bitmovin/analytics/utils/HttpClient.kt
package com.bitmovin.analytics.utils

import android.content.Context
import android.util.Log
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException

class HttpClient(private val context: Context, private val client: OkHttpClient) {
    fun post(
        url: String,
        postBody: String?,
        callback: Callback?,
        useSsaiRouting: Boolean = false,
    ) {
        Log.d(TAG, String.format("Posting Analytics JSON: \n%s\n", postBody))

        val urlWithRouting =
            if (useSsaiRouting) {
                val urlBuilder = url.toHttpUrl().newBuilder()
                urlBuilder.addQueryParameter("routingParam", "ssai")
                urlBuilder.build().toString()
            } else {
                url
            }

        val request =
            Request.Builder()
                .url(urlWithRouting)
                .header("Origin", String.format("http://%s", context.packageName))
                .post(postBody.orEmpty().toRequestBody(JSON_CONTENT_TYPE))
                .build()

        client.newCall(request)
            .enqueue(
                object : Callback {
                    override fun onFailure(
                        call: Call,
                        e: IOException,
                    ) {
                        Log.e(TAG, "HTTP Error: ", e)
                        callback?.onFailure(call, e)
                    }

                    @Throws(IOException::class)
                    override fun onResponse(
                        call: Call,
                        response: Response,
                    ) {
                        callback?.onResponse(call, response)
                        response.close()
                    }
                },
            )
    }

    companion object {
        private val JSON_CONTENT_TYPE: MediaType = "application/json; charset=utf-8".toMediaType()
        private const val TAG = "HttpClient"
    }
}
